### DISCUSSION TITLE
File search on `.gitignore` files

### DISCUSSION THREAD

**Discussion by sapirpol** [eclipse-theia/theia#8427] [General] [answered]
VSCode has an option `Use Global Ignore Files` in their settings so I can search for .gitignore files.
I don't see this option in theia.
How can I search for ignored files?

**Comment by vince-fugnitto** [comment]
@sapirpol we do not have support for the preference `Use Global Ignore Files` today, I actually do not see this preference being used often as searching for ignored files is usually done when needed and not always (like this preference controls).

**Search In Workspace**

If you are referring to `search in workspace`, then you can use the search option `include ignored files` (eye icon), so that the search yields results from ignored files.

![siw-ignore](https://user-images.githubusercontent.com/40359487/91173103-4767c300-e6ab-11ea-9868-eeb4e358f3bb.png)

**File Search**

If you are referring to the `file search` (which returns file results to be opened), then you can trigger the command and looking at the placeholder you'll see an option as how to trigger the search for ignored files (in my case <kbd>ctrl</kbd>+<kbd>p</kbd>):

![file-search-ignore](https://user-images.githubusercontent.com/40359487/91173089-42a30f00-e6ab-11ea-9efb-f99e1d66c425.gif)

**Comment by sapirpol** [comment, marked as answer]
Thanks for your response!
I have a VSCode plugin and I'm using VSCode API `vscode.workspace.findFiles`.
Is there any other option I can use to enable searching for ignored files?

**Reply by vince-fugnitto** [reply]
Based on the documentation, perhaps when passing `null` for `exclude`:

> @param exclude  A [glob pattern](#GlobPattern) that defines files and folders to exclude. The glob pattern
>  will be matched against the file paths of resulting matches relative to their workspace. When `undefined` only default excludes will apply, when `null` no excludes will apply.

https://github.com/eclipse-theia/theia/blob/4390a7d6497a46bc5e5d73911cd563c20cd1a5a2/packages/plugin/src/theia.d.ts#L5166-L5181

It is the same case in [vscode](https://code.visualstudio.com/api/references/vscode-api#workspace):

![image](https://user-images.githubusercontent.com/40359487/91189334-5147f100-e6c0-11ea-876b-f0e840e72f42.png)

**Reply by sapirpol** [reply]
I checked my plugin in theia and vscode, it seems that `vscode.workspace.findFiles` API ignores `null` for `exclude` and acts the same as before with regards to .gitignore files.
I also noticed that in vscode the search for ignored files worked - they set `disregardIgnoreFiles` to true and called`searchService.fileSearch`, so they respect files from .gitignore : https://github.com/microsoft/vscode/blob/f74e473238aca7b79c08be761d99a0232838ca4c/src/vs/workbench/api/browser/mainThreadWorkspace.ts#L147
Why does theia act differently from vscode? Should I open a bug?

**Reply by vince-fugnitto** [reply]
I quickly made an extension to test the feature, in vscode if `exclude: null` is provided then searching in ignored files is supported (such as `node_modules`). It looks like Theia does not support this yet as there is a discrepancy in the results. Please feel free to provide a pull-request to resolve the issue:

- extension: https://github.com/vince-fugnitto/vscode-find-files/releases/download/1.0.0/vscode-find-files-0.0.1.vsix

_VS Code_:

<img width="477" alt="Screen Shot 2020-08-31 at 12 37 57 PM" src="https://user-images.githubusercontent.com/40359487/91744394-725c8600-eb87-11ea-8b04-cab60e3e4132.png">

_Theia_:

<img width="529" alt="Screen Shot 2020-08-31 at 12 43 15 PM" src="https://user-images.githubusercontent.com/40359487/91744464-8dc79100-eb87-11ea-8a48-78c0fdbd76d5.png">
